### DISCUSSION TITLE
OpenTelemetry support in Spring?

### DISCUSSION THREAD

**Discussion by fengli79** [spring-projects/spring-integration#8677] [General]
OpenTelemetry is being more and more popular. It seems Spring is mostly on micrometer. Is there a plan to support OpenTelemetry? Including: log/metrics/tracing

**Comment by artembilan** [comment]
We do support OTel via Micormeter Tracing API: https://micrometer.io/docs/tracing#_micrometer_tracing_opentelemetry_setup.
Spring Boot also exposes respective auto-configuration: https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#actuator.micrometer-tracing.tracers.

This has nothing to do with Spring Integration by itself: we simply rely on the classpath resolution and whenever a Micrometer Observation API delegates to.
You just need to upgrade to the latest Spring stack. Better to trust Spring Boot dependency management: https://spring.io/projects/spring-boot#learn

**Comment by fengli79** [comment]
Thanks @artembilan. The micrometer tracing API looks wrap the OTel trace client side API and provide the full integration already (leverage on all OTel tracing client side offering).
But for metrics, it seems micrometer reports to the backend via OTLP protocol directly without using the OTel client side sdk. In that sense, OTel client side features, such as auto-infer the resource labels on different cloud platforms are not included.

As a library provider, if we want to support open telemetry (in general, including all necessary client side features, not just the telemetry data reporting protocol: otlp), may we get your guidance for the right direction?

For exmaple:
- If we use OTel in our library, while other modules in spring using micrometer, this may introduce some incompatible cases, such as, having different config system for otel and micrometer?
- If we use micrometer, but we still want to get a full stack of otel client side features, can we have a micrometer -> otel sdk wrapper?

The last question is more about the over architect. Is mircrometer is the sole monitoring system/api, which spring to tightly integrated? In another sentence, is Spring open to support other monitoring frameworks (such as OTel) as the first residence, for exmaple, solely relies on OTel to support logging/metrics/tracing without micrometer in the middle as an API broker?

**Reply by artembilan** [reply]
Your questions have nothing to do with Spring Integration Framework: https://spring.io/projects/spring-integration.
Please, seek answers exactly on Micrometer side: https://micrometer.io/.
The Micrometer teams works closely with us, so they can help with Spring-related questions on Micrometer side.
